pkg/operators/filter/filter.go

```diff
@@ -0,0 +1,403 @@
+				case "!=":
+					op = comparisonTypeMatch
+					negate = true
+				case "<=":
+					op = comparisonTypeLte
+				case "<":
+					op = comparisonTypeLt
+				case ">=":
+					op = comparisonTypeGte
+				case ">":
+					op = comparisonTypeGt
+				case "~":
+					op = comparisonTypeRegex
+				default:
+					return fmt.Errorf("invalid operation: %q", opString)
+				}
+				stage = 2
+			}
+		case 2:
+			value += string(filter[pos])
+			pos++
+		}
+	}
+
+	var filterds datasource.DataSource
+	var field datasource.FieldAccessor
+	for _, ds := range gadgetCtx.GetDataSources() {
+		if dsName != "" && ds.Name() != dsName {
+			continue
+		}
+		nf := ds.GetField(fieldName)
+		if nf == nil {
+			continue
+		}
+		if field != nil {
+			return fmt.Errorf("ambiguous field name, please specify the datasource")
+		}
+		field = nf
+		filterds = ds
+	}
+
+	if field == nil {
+		return fmt.Errorf("field %q not found", fieldName)
+	}
+
+	ff, err := getFilterFunc(field, op, negate, value)
+	if err != nil {
+		return err
+	}
+
+	f.ffns[filterds] = append(f.ffns[filterds], ff)
+	return nil
+}
+
+func getFilterFunc(f datasource.FieldAccessor, op comparisonType, negate bool, stringVal string) (func(datasource.DataSource, datasource.Data) bool, error) {
+	var fn func(datasource.DataSource, datasource.Data) bool
+	var intVal int64
+	var uintVal uint64
+	var floatVal float64
+	var boolVal bool
+	var err error
+
+	fieldType := f.Type()
+
+	if fieldType == api.Kind_String && op == comparisonTypeRegex {
+		return nil, fmt.Errorf("regex based filtering can only be used for string types")
+	}
+	if fieldType == api.Kind_Bool && op != comparisonTypeMatch {
+		return nil, fmt.Errorf("boolean values can only be filtered by exact match")
+	}
+
+	bitSize := 64
+	switch f.Type() {
+	case api.Kind_Int8, api.Kind_Uint8:
+		bitSize = 8
+	case api.Kind_Int16, api.Kind_Uint16:
+		bitSize = 16
+	case api.Kind_Int32, api.Kind_Uint32, api.Kind_Float32:
+		bitSize = 32
+	}
+
+	switch f.Type() {
+	default:
+		return nil, fmt.Errorf("unsupported field type for comparison: %s", f.Type())
+	case api.Kind_Int8, api.Kind_Int16, api.Kind_Int32, api.Kind_Int64:
+		intVal, err = strconv.ParseInt(stringVal, 10, bitSize)
+		if err != nil {
+			return nil, fmt.Errorf("parsing comparison value as int: %w", err)
+		}
+	case api.Kind_Uint8, api.Kind_Uint16, api.Kind_Uint32, api.Kind_Uint64:
+		uintVal, err = strconv.ParseUint(stringVal, 10, bitSize)
+		if err != nil {
+			return nil, fmt.Errorf("parsing comparison value as uint: %w", err)
+		}
+	case api.Kind_String, api.Kind_Invalid:
+	// Nothing to be done in this case
+	case api.Kind_Bool:
+		switch stringVal {
+		default:
+			return nil, fmt.Errorf("parsing comparison value %q as bool", stringVal)
+		case "true", "1":
+			boolVal = true
+		case "false", "0":
+		}
+	}
+
+	switch f.Type() {
+	case api.Kind_String, api.Kind_CString:
+		cmp := getCompareFunc[string](op)
+		fn = func(ds datasource.DataSource, data datasource.Data) bool {
```

Comment:
nit: return the function directly in the switch-case block. The `fn` variable won't be needed on the function level scope and the `if` (but not its body) at the end of the function can be removed